drivers/flash/spi_flash_mx25.c

```diff
@@ -0,0 +1,362 @@
+/*
+ * Copyright (c) 2016 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+
+#include <flash.h>
+#include <spi.h>
+#include <init.h>
+#include <string.h>
+#include "spi_flash_mx25_defs.h"
+#include "spi_flash_mx25.h"
+#include "flash_priv.h"
+
+static int spi_flash_wb_access(struct spi_flash_data *ctx,
+			       u8_t cmd, bool addressed, off_t offset,
+			       void *data, size_t length, bool write)
+{
+	u8_t access[4];
+	struct spi_buf buf[2] = {
+		{
+			.buf = access
+		},
+		{
+			.buf = data,
+			.len = length
+		}
+	};
+	struct spi_buf_set tx = {
+		.buffers = buf,
+	};
+
+	access[0] = cmd;
+
+	if (addressed) {
+		access[0] = (u8_t) (offset >> 16);
+		access[1] = (u8_t) (offset >> 8);
+		access[2] = (u8_t) offset;
+		access[3] = cmd;
+
+		buf[0].len = 4;
+		ctx->spi_cfg.operation = SPI_WORD_SET(32);
+	} else {
+		buf[0].len = 1;
+		ctx->spi_cfg.operation = SPI_WORD_SET(8);
+	}
+
+	tx.count = length ? 2 : 1;
+
+	if (!write) {
+		const struct spi_buf_set rx = {
+			.buffers = (buf + 1),
+			.count = 2
+		};
+		tx.count = 1;
+		return spi_transceive(ctx->spi, &ctx->spi_cfg, &tx, &rx);
+	}
+
+	return spi_write(ctx->spi, &ctx->spi_cfg, &tx);
+}
+
+static inline int spi_flash_wb_id(struct device *dev)
+{
+	struct spi_flash_data *const driver_data = dev->driver_data;
+	u32_t temp_data;
+	u8_t buf[5], i;
+
+	if (spi_flash_wb_access(driver_data, MX25_CMD_RDID,
+				false, 0, buf, 3, false) != 0) {
+		return -EIO;
+	}
+
+	temp_data = ((u32_t) buf[0]) << 16;
+	temp_data |= ((u32_t) buf[1]) << 8;
+	temp_data |= (u32_t) buf[2];
+
+	if (temp_data != MX25_RDID_VALUE) {
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+static u8_t spi_flash_wb_reg_read(struct device *dev, u8_t reg)
+{
+	struct spi_flash_data *const driver_data = dev->driver_data;
+
+	if (spi_flash_wb_access(driver_data, reg,
+				false, 0, &reg, 1, false)) {
+		return 0;
+	}
+
+	return reg;
+}
+
+static inline void wait_for_flash_idle1(struct device *dev)
+{
+	u8_t reg;
+	struct spi_flash_data *data = dev->driver_data;
+
+	do {
+		reg = spi_flash_wb_reg_read(dev, MX25_CMD_RDSR);
+	} while (reg & MX25_WEL_BIT);
```

Comment:
would be nice to have a safe-guard loop counter here. It could possibly loop forever if the hw has some malfunction.